### PR TITLE
feat(mito2): concat and projection

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?
- Use `arrow::compute::concat` to concat vectors in memtable
- Respect projection

## Checklist

- [X]  I have written the necessary rustdoc comments.
- [X]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/mito2/src/memtable/time_series.rs [hunk 1]
```diff
@@ -339,29 +360,112 @@ struct Values {
...
+    }
+
     /// Converts [Values] to `Batch`, sorts the batch according to `timestamp, sequence` desc and
     /// keeps only the latest row for the same timestamp.
-    pub fn to_batch(&self, primary_key: &[u8], metadata: &RegionMetadataRef) -> Result<Batch> {
+    pub fn to_batch(
+        &self,
+        primary_key: &[u8],
+        metadata: &RegionMetadataRef,
+        projection: Option<&[ColumnId]>,
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:390]: Could we just store the projection as a `HashSet<ColumnId>` and check whether we need a column by searching the hash set?

File: src/mito2/src/memtable/time_series.rs [hunk 2]
```diff
@@ -237,39 +253,44 @@ impl Series {
         self.freeze(region_metadata);
 
         let mut frozen = self.frozen.clone();
-        let values = if frozen.len() == 1 {
+        let values = if frozen.is_empty() {
+            Values::new_empty(region_metadata)
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:251]: Is it possible that frozen is empty? A series in the series set should at least contain one point. I think we could also add an assertion.

File: src/mito2/src/memtable/time_series.rs [hunk 3]
```diff
@@ -339,29 +354,112 @@ struct Values {
...
+    fn from_columns(cols: &[ArrayRef]) -> Result<Self> {
+        debug_assert!(cols.len() >= 3);
+        let timestamp =
+            Helper::try_into_vector(cols.get(0).unwrap()).context(ConvertVectorSnafu)?;
+        let sequence = Arc::new(
+            UInt64Vector::try_from_arrow_array(cols.get(1).unwrap()).context(ConvertVectorSnafu)?,
+        );
+        let op_type = Arc::new(
+            UInt8Vector::try_from_arrow_array(cols.get(2).unwrap()).context(ConvertVectorSnafu)?,
+        );
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:453]:
```suggestion
        let timestamp =
            Helper::try_into_vector(&cols[0]).context(ConvertVectorSnafu)?;
        let sequence = Arc::new(
            UInt64Vector::try_from_arrow_array(&cols[1]).context(ConvertVectorSnafu)?,
        );
        let op_type = Arc::new(
            UInt8Vector::try_from_arrow_array(&cols[2]).context(ConvertVectorSnafu)?,
        );
```

File: src/mito2/src/memtable/time_series.rs [hunk 4]
```diff
@@ -237,39 +247,44 @@ impl Series {
...
-                    .op_type
-                    .extend_slice_of(&*v.op_type, 0, len)
-                    .context(CompactValuesSnafu)?;
-
-                for (idx, f) in v.fields.iter().enumerate() {
-                    builder.fields[idx]
-                        .extend_slice_of(&**f, 0, len)
-                        .context(CompactValuesSnafu)?;
-                }
+            let column_size = frozen.get(0).unwrap().fields.len() + 3;
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:258]: Same as `get(0).unwrap()`
```suggestion
            let column_size = frozen[0].fields.len() + 3;
```

File: src/mito2/src/memtable/time_series.rs [hunk 5]
```diff
@@ -237,39 +247,44 @@ impl Series {
...
+                })
+                .collect::<Vec<_>>();
+
+            let concatenated = cols
+                .into_iter()
+                .map(|col| {
+                    arrow::compute::concat(&col.iter().map(|a| a.as_ref()).collect::<Vec<_>>())
+                })
+                .collect::<std::result::Result<Vec<_>, _>>()
+                .context(ComputeArrowSnafu)?;
```
Comment (evenyag) [inline src/mito2/src/memtable/time_series.rs:284]: Could we return a `Vec<ArrayRef>` in `iter_columns()`? Maybe rename it to `columns()`
```rust
let mut arrays_vec = frozen.iter().map(|v| v.columns()).collect::<Vec<_>>();
let concatenated = (0..column_size)
    .map(|i| {
        // safety: every `Values` has same column num.
        let to_concat = arrays_vec.iter().map(|arrays| arrays[i].as_ref()).collect::<Vec<_>>();
        arrow::compute::concat(&to_concat)
    })
    .collect::<std::result::Result<Vec<_>, _>>()
    .context(ComputeArrowSnafu)?;
```

File: src/mito2/src/memtable/time_series.rs (hunk 1 above)
Comment (v0y4g3r) [reply src/mito2/src/memtable/time_series.rs:390]: I think we should preserve the order of field columns as in projection.

File: src/mito2/src/memtable/time_series.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/mito2/src/memtable/time_series.rs]: No, given that we cannot delete rows from memtable, each series must contain at least one row and hence one frozen value.

**Review by evenyag — APPROVED** [review]

**Review by waynexia — APPROVED** [review]
LGTM

**Pull request merged**